milli/src/search/new/graph_based_ranking_rule.rs

```diff
@@ -0,0 +1,391 @@
+            }
+        } else {
+            query_graph.nodes.map(|_| None)
+        };
+
+        let graph = RankingRuleGraph::build(ctx, query_graph.clone(), removal_cost)?;
+        let condition_docids_cache = ConditionDocIdsCache::default();
+        let dead_ends_cache = DeadEndsCache::new(&graph.conditions_interner);
+
+        // Then pre-compute the cost of all paths from each node to the end node
+        let all_costs = graph.find_all_costs_to_end();
+
+        let state = GraphBasedRankingRuleState {
+            graph,
+            conditions_cache: condition_docids_cache,
+            dead_ends_cache,
+            all_costs,
+            cur_cost: 0,
+        };
+
+        self.state = Some(state);
+
+        Ok(())
+    }
+
+    fn next_bucket(
+        &mut self,
+        ctx: &mut SearchContext<'ctx>,
+        logger: &mut dyn SearchLogger<QueryGraph>,
+        universe: &RoaringBitmap,
+    ) -> Result<Option<RankingRuleOutput<QueryGraph>>> {
+        // If universe.len() <= 1, the bucket sort algorithm
+        // should not have called this function.
+        assert!(universe.len() > 1);
+        // Will crash if `next_bucket` is called before `start_iteration` or after `end_iteration`,
+        // should never happen
+        let mut state = self.state.take().unwrap();
+
+        // Retrieve the cost of the paths to compute
+        let Some(&cost) = state
+            .all_costs
+            .get(state.graph.query_graph.root_node)
+            .iter()
+            .find(|c| **c >= state.cur_cost) else {
+                self.state = None;
+                return Ok(None);
+        };
+        state.cur_cost = cost + 1;
+
+        let mut bucket = RoaringBitmap::new();
+
+        let GraphBasedRankingRuleState {
+            graph,
+            conditions_cache: condition_docids_cache,
+            dead_ends_cache,
+            all_costs,
+            cur_cost: _,
+        } = &mut state;
+
+        let mut universe = universe.clone();
+
+        let mut used_conditions = SmallBitmap::for_interned_values_in(&graph.conditions_interner);
+        let mut good_paths = vec![];
+        let mut considered_paths = vec![];
+
+        // For each path of the given cost, we will compute its associated
+        // document ids.
+        // In case the path does not resolve to any document id, we try to figure out why
+        // and update the `dead_ends_cache` accordingly.
+        // Updating the dead_ends_cache helps speed up the execution of `visit_paths_of_cost` and reduces
+        // the number of future candidate paths given by that same function.
+
+        let mut subpaths_docids: Vec<(Interned<G::Condition>, RoaringBitmap)> = vec![];
+
+        let mut nodes_with_removed_outgoing_conditions = BTreeSet::new();
+        let visitor = PathVisitor::new(cost, graph, all_costs, dead_ends_cache);
+
+        visitor.visit_paths(&mut |path, graph, dead_ends_cache| {
```

Comment:
Explanation: the role of this closure is to:
1. Compute the docids of the path
2. Say whether we should continue looking for more paths or not (we shouldn't if we have exhaustively sorted the whole universe)
3. Update the `dead_ends_cache` in case the path yields no docids to speed up the path finding algorithm and avoid computing the docids of paths that are similar (i.e. share same prefix that yields no docids).